src/wrappers/themis/react-native-themis/ios/RCTThemis.m

```diff
@@ -50,131 +50,154 @@ + (BOOL)requiresMainQueueSetup
 
 + (NSArray*)dataSerialize:(NSData*) data
 {
-  if (data == nil) return nil;
-  
-  const char* buffer = (const char*) data.bytes;
-  NSMutableArray<NSNumber*> *array = [[NSMutableArray alloc] init];
-  
-  for (NSInteger i = 0; i < data.length; i++) {
-    NSNumber *num = [[NSNumber alloc] initWithUnsignedChar: buffer[i]];
-    [array addObject:num];
-  }
-  return [array copy];
+    if (data == nil || data.length == 0 ) return nil;
+
+    const char* buffer = (const char*) data.bytes;
+    NSMutableArray<NSNumber*> *array = [[NSMutableArray alloc] init];
+
+    for (NSInteger i = 0; i < data.length; i++) {
+        NSNumber *num = [[NSNumber alloc] initWithUnsignedChar: buffer[i]];
+        [array addObject:num];
+    }
+    return [array copy];
 }
 
 /*************************************************/
 /* Used to import NSData* from React Native app  */
 /* NSData <-- NSArray[NSNumber(unsigned char)]   */
 /*                                               */
 
-
 + (NSData*)dataDeserialize:(NSArray*) data
 {
-  if (data == nil) return nil;
-  
-  char* buffer = (char*)malloc(data.count);
-  if (buffer == nil) {
-    return nil; // malloc returns nil
-  }
-  
-  for (NSInteger i = 0; i < data.count; i++) {
-    NSNumber *num = data[i];
-    const unsigned char c = num.unsignedCharValue;
-    if (c < 0 || c > 255) {
-      NSException *e = [NSException
-                        exceptionWithName:@"ByteOverflowException"
-                        reason:@"Value is out of range"
-                        userInfo:nil];
-      @throw e;
-    }
-    buffer[i] = c;
-  }
+    if ( data == nil || data.count == 0 ) {
+        NSException *exception = [NSException
+                                  exceptionWithName:@"DeserializationError: empty input data"
+                                  reason:@DESERIALIZE_ERRORREASON
+                                  userInfo:@{@"errorCode": @DESERIALIZE_ERROR}
+        ];
+        @throw exception;
+    }
 
-  NSData *result = [NSData dataWithBytesNoCopy:buffer length:data.count freeWhenDone:YES];
-  
-  return result;
+    char* buffer = (char*)malloc(data.count);
+
+    if (buffer == nil) {
+        NSException *exception = [NSException
+                                  exceptionWithName:@"DeserializationError: can not allocate memory"
+                                  reason:@DESERIALIZE_ERRORREASON
+                                  userInfo:@{@"errorCode": @DESERIALIZE_ERROR}
+        ];
+        @throw exception;
+    }
+
+    for (NSInteger i = 0; i < data.count; i++) {
+        NSNumber *num = data[i];
+        const unsigned char c = num.unsignedCharValue;
+        if (c < 0 || c > 255) {
+            NSException *e = [NSException
```

Comment:
do I understand correctly that we check that value of unsigned char type is in range of 0..255? is there any case when it can be?

Reply:
I believe this check is misplaced and it should be done before unsigned char conversion. NSNumber can indeed hold 32-bit values.

Reply:
Ok, I will check num.intValue to be in range 0...255.

Reply:
> Ok, I will check num.intValue to be in range 0...255.

It's better to use the `compare` method, if you go full paranoid. For all I know, I would totally expect JS runtime to put literal `doubles` into NSNumber. (You can check with `objCType` what it holds.)

Reply:
used compare method.